File: backend/app/services/admin_user_management_service.py
from datetime import datetime
from datetime import timezone
from typing import Any

from fastapi import HTTPException
from fastapi import status
from sqlalchemy.orm import Session

from app.enums import AuthProvider
from app.enums import Role
from app.repositories.user_repository import (
    UserRepository,
)
from app.repositories.visitor_log_repository import (
    VisitorLogRepository,
)

from app.models.user import User


class AdminUserManagementService:

    def __init__(
        self,
        db: Session,
    ) -> None:

        self.db = db

        self.user_repository = (
            UserRepository(db)
        )

        self.visitor_log_repository = (
            VisitorLogRepository(db)
        )

    # =====================================================
    # GET REGISTERED USERS
    # =====================================================

    def get_registered_users(
        self,
        skip: int = 0,
        limit: int = 25,
        search: str | None = None,
        is_active: bool | None = None,
        is_locked: bool | None = None,
        is_verified: bool | None = None,
        auth_provider: AuthProvider | None = None,
    ) -> dict[str, Any]:

        safe_skip = max(
            int(skip or 0),
            0,
        )

        safe_limit = min(
            max(
                int(limit or 25),
                1,
            ),
            100,
        )

        users = (
            self.user_repository
            .get_all(
                skip=safe_skip,
                limit=safe_limit,
                search=search,
                is_active=is_active,
                is_locked=is_locked,
                is_verified=is_verified,
                auth_provider=auth_provider,
                include_deleted=False,
                registered_only=True,
            )
        )

        total = (
            self.user_repository
            .count_all(
                search=search,
                is_active=is_active,
                is_locked=is_locked,
                is_verified=is_verified,
                auth_provider=auth_provider,
                include_deleted=False,
                registered_only=True,
            )
        )

        return {
            "total": int(
                total or 0
            ),
            "skip": safe_skip,
            "limit": safe_limit,
            "users": users,
        }

    # =====================================================
    # GET REGISTERED USER DETAIL
    # =====================================================

    def get_registered_user(
        self,
        user_id: int,
    ) -> User:

        user = self.user_repository.get_by_id(
            user_id=user_id,
            include_deleted=False,
        )

        if user is None:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Registered user not found.",
            )

        if user.role != Role.REGISTERED:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Registered user not found.",
            )

        return user

    # =====================================================
    # GET USER SUMMARY
    # =====================================================

    def get_registered_user_summary(
        self,
    ) -> dict[str, int]:

        total = (
            self.user_repository
            .count_all(
                registered_only=True,
            )
            or 0
        )

        active = (
            self.user_repository
            .count_all(
                is_active=True,
                registered_only=True,
            )
            or 0
        )

        inactive = (
            self.user_repository
            .count_all(
                is_active=False,
                registered_only=True,
            )
            or 0
        )

        locked = (
            self.user_repository
            .count_all(
                is_locked=True,
                registered_only=True,
            )
            or 0
        )

        verified = (
            self.user_repository
            .count_all(
                is_verified=True,
                registered_only=True,
            )
            or 0
        )

        local_accounts = (
            self.user_repository
            .count_all(
                auth_provider=(
                    AuthProvider.LOCAL
                ),
                registered_only=True,
            )
            or 0
        )

        google_accounts = (
            self.user_repository
            .count_all(
                auth_provider=(
                    AuthProvider.GOOGLE
                ),
                registered_only=True,
            )
            or 0
        )

        return {
            "total": int(total),
            "active": int(active),
            "inactive": int(inactive),
            "locked": int(locked),
            "verified": int(verified),
            "local_accounts": int(
                local_accounts
            ),
            "google_accounts": int(
                google_accounts
            ),
        }

    # =====================================================
    # ACTIVATE USER
    # =====================================================

    def activate_user(
        self,
        user_id: int,
    ) -> dict[str, Any]:

        user = self.get_registered_user(
            user_id
        )

        if user.is_active:
            return {
                "user_id": user.id,
                "action": "ACTIVATE",
                "is_active": True,
                "is_locked": bool(
                    user.is_locked
                ),
                "message": (
                    "The registered user is already active."
                ),
            }

        updated_user = (
            self.user_repository
            .activate(user)
        )

        return {
            "user_id": updated_user.id,
            "action": "ACTIVATE",
            "is_active": bool(
                updated_user.is_active
            ),
            "is_locked": bool(
                updated_user.is_locked
            ),
            "message": (
                "Registered user activated successfully."
            ),
        }

    # =====================================================
    # DEACTIVATE USER
    # =====================================================

    def deactivate_user(
        self,
        user_id: int,
    ) -> dict[str, Any]:

        user = self.get_registered_user(
            user_id
        )

        if not user.is_active:
            return {
                "user_id": user.id,
                "action": "DEACTIVATE",
                "is_active": False,
                "is_locked": bool(
                    user.is_locked
                ),
                "message": (
                    "The registered user is already inactive."
                ),
            }

        updated_user = (
            self.user_repository
            .deactivate(user)
        )

        return {
            "user_id": updated_user.id,
            "action": "DEACTIVATE",
            "is_active": bool(
                updated_user.is_active
            ),
            "is_locked": bool(
                updated_user.is_locked
            ),
            "message": (
                "Registered user deactivated successfully."
            ),
        }

    # =====================================================
    # LOCK USER
    # =====================================================

    def lock_user(
        self,
        user_id: int,
    ) -> dict[str, Any]:

        user = self.get_registered_user(
            user_id
        )

        if user.is_locked:
            return {
                "user_id": user.id,
                "action": "LOCK",
                "is_active": bool(
                    user.is_active
                ),
                "is_locked": True,
                "message": (
                    "The registered user is already locked."
                ),
            }

        updated_user = (
            self.user_repository
            .lock(user)
        )

        return {
            "user_id": updated_user.id,
            "action": "LOCK",
            "is_active": bool(
                updated_user.is_active
            ),
            "is_locked": bool(
                updated_user.is_locked
            ),
            "message": (
                "Registered user locked successfully."
            ),
        }

    # =====================================================
    # UNLOCK USER
    # =====================================================

    def unlock_user(
        self,
        user_id: int,
    ) -> dict[str, Any]:

        user = self.get_registered_user(
            user_id
        )

        if not user.is_locked:
            return {
                "user_id": user.id,
                "action": "UNLOCK",
                "is_active": bool(
                    user.is_active
                ),
                "is_locked": False,
                "message": (
                    "The registered user is already unlocked."
                ),
            }

        updated_user = (
            self.user_repository
            .unlock(user)
        )

        return {
            "user_id": updated_user.id,
            "action": "UNLOCK",
            "is_active": bool(
                updated_user.is_active
            ),
            "is_locked": bool(
                updated_user.is_locked
            ),
            "message": (
                "Registered user unlocked successfully."
            ),
        }

    # =====================================================
    # SOFT DELETE USER
    # =====================================================

    def delete_user(
        self,
        user_id: int,
    ) -> dict[str, Any]:

        user = self.get_registered_user(
            user_id
        )

        deleted_user = (
            self.user_repository
            .soft_delete(
                user=user,
                deleted_at=datetime.now(
                    timezone.utc
                ),
            )
        )

        return {
            "user_id": deleted_user.id,
            "deleted": (
                deleted_user.deleted_at
                is not None
            ),
            "message": (
                "Registered user deleted successfully."
            ),
        }

    # =====================================================
    # GET GUEST VISITOR LOGS
    # =====================================================

    def get_guest_visitor_logs(
        self,
        session_id: str | None = None,
        browser: str | None = None,
        operating_system: str | None = None,
        visited_page: str | None = None,
        date_from: datetime | None = None,
        date_to: datetime | None = None,
        skip: int = 0,
        limit: int = 25,
    ) -> dict[str, Any]:

        safe_skip = max(
            int(skip or 0),
            0,
        )

        safe_limit = min(
            max(
                int(limit or 25),
                1,
            ),
            100,
        )

        logs = (
            self.visitor_log_repository
            .get_all(
                session_id=session_id,
                browser=browser,
                operating_system=(
                    operating_system
                ),
                visited_page=visited_page,
                date_from=date_from,
                date_to=date_to,
                guest_only=True,
                skip=safe_skip,
                limit=safe_limit,
            )
        )

        total = (
            self.visitor_log_repository
            .count_filtered(
                session_id=session_id,
                browser=browser,
                operating_system=(
                    operating_system
                ),
                visited_page=visited_page,
                date_from=date_from,
                date_to=date_to,
                guest_only=True,
            )
        )

        return {
            "total": int(
                total or 0
            ),
            "skip": safe_skip,
            "limit": safe_limit,
            "logs": logs,
        }

    # =====================================================
    # GET GUEST SUMMARY
    # =====================================================

    def get_guest_visitor_summary(
        self,
    ) -> dict[str, Any]:

        top_pages = (
            self.visitor_log_repository
            .top_pages(
                limit=5,
                guest_only=True,
            )
        )

        top_browsers = (
            self.visitor_log_repository
            .top_browsers(
                limit=5,
                guest_only=True,
            )
        )

        top_operating_systems = (
            self.visitor_log_repository
            .top_operating_systems(
                limit=5,
                guest_only=True,
            )
        )

        return {
            "total_guest_visits": int(
                self.visitor_log_repository
                .count_guest_visits()
                or 0
            ),

            "unique_guest_visitors": int(
                self.visitor_log_repository
                .count_unique_guest_visitors()
                or 0
            ),

            "today_guest_visits": int(
                self.visitor_log_repository
                .count_today_guest_visits()
                or 0
            ),

            "today_unique_guest_visitors": int(
                self.visitor_log_repository
                .count_today_unique_guest_visitors()
                or 0
            ),

            "top_pages": (
                self._normalize_analytics_rows(
                    rows=top_pages,
                    label_keys=[
                        "visited_page",
                        "page",
                    ],
                )
            ),

            "top_browsers": (
                self._normalize_analytics_rows(
                    rows=top_browsers,
                    label_keys=[
                        "browser",
                    ],
                )
            ),

            "top_operating_systems": (
                self._normalize_analytics_rows(
                    rows=(
                        top_operating_systems
                    ),
                    label_keys=[
                        "operating_system",
                    ],
                )
            ),
        }

    # =====================================================
    # GET COMPLETE MANAGEMENT SUMMARY
    # =====================================================

    def get_management_summary(
        self,
    ) -> dict[str, Any]:

        return {
            "registered_users": (
                self.get_registered_user_summary()
            ),
            "guest_visitors": (
                self.get_guest_visitor_summary()
            ),
        }

    # =====================================================
    # NORMALIZE ANALYTICS ROWS
    # =====================================================

    @staticmethod
    def _normalize_analytics_rows(
        rows,
        label_keys: list[str],
    ) -> list[dict[str, Any]]:

        results: list[
            dict[str, Any]
        ] = []

        for row in rows or []:

            label = None
            visits = 0

            if isinstance(
                row,
                dict,
            ):
                for key in label_keys:
                    if row.get(key):
                        label = row.get(key)
                        break

                visits = (
                    row.get("visit_count")
                    or row.get("visits")
                    or row.get("count")
                    or 0
                )

            elif hasattr(
                row,
                "_mapping",
            ):
                mapping = row._mapping

                for key in label_keys:
                    if mapping.get(key):
                        label = mapping.get(
                            key
                        )
                        break

                visits = (
                    mapping.get(
                        "visit_count"
                    )
                    or mapping.get(
                        "visits"
                    )
                    or mapping.get(
                        "count"
                    )
                    or 0
                )

            elif isinstance(
                row,
                (
                    tuple,
                    list,
                ),
            ):
                if len(row) >= 1:
                    label = row[0]

                if len(row) >= 2:
                    visits = row[1]

            else:
                for key in label_keys:
                    value = getattr(
                        row,
                        key,
                        None,
                    )

                    if value:
                        label = value
                        break

                visits = (
                    getattr(
                        row,
                        "visit_count",
                        None,
                    )
                    or getattr(
                        row,
                        "visits",
                        None,
                    )
                    or getattr(
                        row,
                        "count",
                        None,
                    )
                    or 0
                )

            results.append(
                {
                    "label": str(
                        label
                        or "Unknown"
                    ),
                    "visits": int(
                        visits or 0
                    ),
                }
            )

        return results

File: frontend/src/services/adminUserManagementService.js
import axios from "axios";

/* =====================================================
   API CONFIGURATION
===================================================== */

const API_URL =
  import.meta.env.VITE_API_URL ||
  import.meta.env.VITE_API_BASE_URL ||
  "http://127.0.0.1:8000";

const USER_MANAGEMENT_ENDPOINT = "/api/admin/user-management";

/*
 * Change this only if Swagger shows a different path.
 *
 * Examples:
 * "/refresh"
 * "/api/auth/refresh"
 * "/api/admin/auth/refresh"
 */
const REFRESH_ENDPOINT = "/api/auth/refresh";

/* =====================================================
   AUTHENTICATED API
===================================================== */

const adminUserManagementAPI = axios.create({
  baseURL: API_URL,

  timeout: 120000,

  headers: {
    Accept: "application/json",
    "Content-Type": "application/json",
  },
});

const refreshAPI = axios.create({
  baseURL: API_URL,

  timeout: 30000,

  headers: {
    Accept: "application/json",
    "Content-Type": "application/json",
  },
});

let refreshPromise = null;

/* =====================================================
   TOKEN HELPERS
===================================================== */

function getAccessToken() {
  return (
    localStorage.getItem("admin_token") ||
    localStorage.getItem("admin_access_token")
  );
}

function getRefreshToken() {
  return localStorage.getItem("admin_refresh_token");
}

function saveRefreshedTokens(data) {
  const accessToken = data?.access_token;

  if (!accessToken) {
    throw new Error("The refresh response did not include an access token.");
  }

  localStorage.setItem("admin_token", accessToken);

  localStorage.setItem("admin_access_token", accessToken);

  if (data?.refresh_token) {
    localStorage.setItem("admin_refresh_token", data.refresh_token);
  }

  return accessToken;
}

/* =====================================================
   CLEAR SESSION
===================================================== */

export function clearAdminUserManagementSession() {
  localStorage.removeItem("admin_token");

  localStorage.removeItem("admin_access_token");

  localStorage.removeItem("admin_refresh_token");

  localStorage.removeItem("admin_user");
}

function redirectToAdminLogin() {
  if (window.location.pathname !== "/admin/login") {
    window.location.replace("/admin/login");
  }
}

/* =====================================================
   REQUEST INTERCEPTOR
===================================================== */

adminUserManagementAPI.interceptors.request.use(
  (config) => {
    const accessToken = getAccessToken();

    config.headers = config.headers || {};

    if (accessToken) {
      config.headers.Authorization = `Bearer ${accessToken}`;
    }

    config.headers.Accept = config.headers.Accept || "application/json";

    return config;
  },

  (error) => Promise.reject(error),
);

/* =====================================================
   RESPONSE INTERCEPTOR
===================================================== */

adminUserManagementAPI.interceptors.response.use(
  (response) => response,

  async (error) => {
    const originalRequest = error?.config;

    const responseStatus = error?.response?.status;

    if (!originalRequest) {
      return Promise.reject(error);
    }

    const requestUrl = String(originalRequest.url || "");

    const isRefreshRequest = requestUrl.includes(REFRESH_ENDPOINT);

    if (responseStatus !== 401 || originalRequest._retry || isRefreshRequest) {
      return Promise.reject(error);
    }

    const refreshToken = getRefreshToken();

    if (!refreshToken) {
      clearAdminUserManagementSession();

      redirectToAdminLogin();

      return Promise.reject(error);
    }

    originalRequest._retry = true;

    try {
      if (!refreshPromise) {
        refreshPromise = refreshAPI
          .post(REFRESH_ENDPOINT, {
            refresh_token: refreshToken,
          })
          .then((response) => saveRefreshedTokens(response?.data || {}))
          .finally(() => {
            refreshPromise = null;
          });
      }

      const newAccessToken = await refreshPromise;

      originalRequest.headers = originalRequest.headers || {};

      originalRequest.headers.Authorization = `Bearer ${newAccessToken}`;

      return adminUserManagementAPI(originalRequest);
    } catch (refreshError) {
      clearAdminUserManagementSession();

      redirectToAdminLogin();

      return Promise.reject(refreshError);
    }
  },
);

/* =====================================================
   ERROR MESSAGE
===================================================== */

export function getAdminUserManagementErrorMessage(error) {
  const detail = error?.response?.data?.detail;

  if (typeof detail === "string") {
    return detail;
  }

  if (Array.isArray(detail)) {
    return detail
      .map((item) => {
        if (typeof item === "string") {
          return item;
        }

        return item?.msg || item?.message || "Invalid request value.";
      })
      .join(", ");
  }

  if (detail && typeof detail === "object") {
    return detail.message || detail.error || "The server rejected the request.";
  }

  switch (error?.response?.status) {
    case 400:
      return "The user management request is invalid.";

    case 401:
      return "Your administrator session expired and could not be refreshed.";

    case 403:
      return "You are not authorized to manage registered users or visitor logs.";

    case 404:
      return "The requested registered user or visitor record was not found.";

    case 409:
      return "The requested action conflicts with the user's current account state.";

    case 422:
      return "One or more user management request values are invalid.";

    case 500:
      return "The server encountered an error while processing the user management request.";

    default:
      return error?.message || "An unexpected user management error occurred.";
  }
}

/* =====================================================
   NORMALIZATION HELPERS
===================================================== */

function normalizePage(page) {
  return Math.max(Number(page) || 1, 1);
}

function normalizePageSize(pageSize) {
  return Math.min(Math.max(Number(pageSize) || 25, 1), 100);
}

function requireUserId(userId) {
  const normalizedId = Number(userId);

  if (!Number.isInteger(normalizedId) || normalizedId <= 0) {
    throw new Error("A valid registered user ID is required.");
  }

  return normalizedId;
}

/* =====================================================
   BUILD USER PARAMS
===================================================== */

function buildUserParams({
  page = 1,
  pageSize = 25,
  search = "",
  isActive = "",
  isLocked = "",
  isVerified = "",
  authProvider = "",
} = {}) {
  const safePage = normalizePage(page);

  const safePageSize = normalizePageSize(pageSize);

  const params = {
    skip: (safePage - 1) * safePageSize,

    limit: safePageSize,
  };

  const cleanSearch = String(search || "").trim();

  if (cleanSearch) {
    params.search = cleanSearch;
  }

  if (isActive !== "" && isActive !== null && isActive !== undefined) {
    params.is_active = isActive === true || isActive === "true";
  }

  if (isLocked !== "" && isLocked !== null && isLocked !== undefined) {
    params.is_locked = isLocked === true || isLocked === "true";
  }

  if (isVerified !== "" && isVerified !== null && isVerified !== undefined) {
    params.is_verified = isVerified === true || isVerified === "true";
  }

  const normalizedProvider = String(authProvider || "").trim();

  if (normalizedProvider) {
    params.auth_provider = normalizedProvider;
  }

  return params;
}

/* =====================================================
   BUILD GUEST LOG PARAMS
===================================================== */

function buildGuestLogParams({
  page = 1,
  pageSize = 25,
  sessionId = "",
  browser = "",
  operatingSystem = "",
  visitedPage = "",
  dateFrom = "",
  dateTo = "",
} = {}) {
  const safePage = normalizePage(page);

  const safePageSize = normalizePageSize(pageSize);

  const params = {
    skip: (safePage - 1) * safePageSize,

    limit: safePageSize,
  };

  const cleanSessionId = String(sessionId || "").trim();

  if (cleanSessionId) {
    params.session_id = cleanSessionId;
  }

  const cleanBrowser = String(browser || "").trim();

  if (cleanBrowser) {
    params.browser = cleanBrowser;
  }

  const cleanOperatingSystem = String(operatingSystem || "").trim();

  if (cleanOperatingSystem) {
    params.operating_system = cleanOperatingSystem;
  }

  const cleanVisitedPage = String(visitedPage || "").trim();

  if (cleanVisitedPage) {
    params.visited_page = cleanVisitedPage;
  }

  if (dateFrom) {
    params.date_from = dateFrom;
  }

  if (dateTo) {
    params.date_to = dateTo;
  }

  return params;
}

/* =====================================================
   GET MANAGEMENT SUMMARY
===================================================== */

export async function getAdminUserManagementSummary() {
  const response = await adminUserManagementAPI.get(
    `${USER_MANAGEMENT_ENDPOINT}/summary`,
  );

  return (
    response?.data || {
      registered_users: {
        total: 0,
        active: 0,
        inactive: 0,
        locked: 0,
        verified: 0,
        local_accounts: 0,
        google_accounts: 0,
      },

      guest_visitors: {
        total_guest_visits: 0,
        unique_guest_visitors: 0,
        today_guest_visits: 0,
        today_unique_guest_visitors: 0,
        top_pages: [],
        top_browsers: [],
        top_operating_systems: [],
      },
    }
  );
}

/* =====================================================
   GET REGISTERED USERS
===================================================== */

export async function getRegisteredUsers({
  page = 1,
  pageSize = 25,
  search = "",
  isActive = "",
  isLocked = "",
  isVerified = "",
  authProvider = "",
} = {}) {
  const params = buildUserParams({
    page,
    pageSize,
    search,
    isActive,
    isLocked,
    isVerified,
    authProvider,
  });

  const response = await adminUserManagementAPI.get(
    `${USER_MANAGEMENT_ENDPOINT}/users`,
    {
      params,
    },
  );

  const data = response?.data || {};

  return {
    total: Number(data.total || 0),

    skip: Number(data.skip || 0),

    limit: Number(data.limit || params.limit),

    users: Array.isArray(data.users) ? data.users : [],
  };
}

/* =====================================================
   GET REGISTERED USER DETAIL
===================================================== */

export async function getRegisteredUserDetail(userId) {
  const parsedUserId = Number(userId);

  if (!Number.isInteger(parsedUserId) || parsedUserId <= 0) {
    throw new Error("A valid user ID is required.");
  }

  const response = await adminUserManagementAPI.get(
    `${USER_MANAGEMENT_ENDPOINT}/users/${parsedUserId}`,
  );

  return response.data;
}

/* =====================================================
   ACTIVATE USER
===================================================== */

export async function activateRegisteredUser(userId) {
  const id = requireUserId(userId);

  const response = await adminUserManagementAPI.post(
    `${USER_MANAGEMENT_ENDPOINT}/users/${id}/activate`,
    {},
  );

  return response.data;
}

/* =====================================================
   DEACTIVATE USER
===================================================== */

export async function deactivateRegisteredUser(userId) {
  const id = requireUserId(userId);

  const response = await adminUserManagementAPI.post(
    `${USER_MANAGEMENT_ENDPOINT}/users/${id}/deactivate`,
    {},
  );

  return response.data;
}

/* =====================================================
   LOCK USER
===================================================== */

export async function lockRegisteredUser(userId) {
  const id = requireUserId(userId);

  const response = await adminUserManagementAPI.post(
    `${USER_MANAGEMENT_ENDPOINT}/users/${id}/lock`,
    {},
  );

  return response.data;
}

/* =====================================================
   UNLOCK USER
===================================================== */

export async function unlockRegisteredUser(userId) {
  const id = requireUserId(userId);

  const response = await adminUserManagementAPI.post(
    `${USER_MANAGEMENT_ENDPOINT}/users/${id}/unlock`,
    {},
  );

  return response.data;
}

/* =====================================================
   DELETE USER
===================================================== */

export async function deleteRegisteredUser(userId) {
  const id = requireUserId(userId);

  const response = await adminUserManagementAPI.delete(
    `${USER_MANAGEMENT_ENDPOINT}/users/${id}`,
  );

  return response.data;
}

/* =====================================================
   GET GUEST VISITOR LOGS
===================================================== */

export async function getGuestVisitorLogs({
  page = 1,
  pageSize = 25,
  sessionId = "",
  browser = "",
  operatingSystem = "",
  visitedPage = "",
  dateFrom = "",
  dateTo = "",
} = {}) {
  const params = buildGuestLogParams({
    page,
    pageSize,
    sessionId,
    browser,
    operatingSystem,
    visitedPage,
    dateFrom,
    dateTo,
  });

  const response = await adminUserManagementAPI.get(
    `${USER_MANAGEMENT_ENDPOINT}/guest-logs`,
    {
      params,
    },
  );

  const data = response?.data || {};

  return {
    total: Number(data.total || 0),

    skip: Number(data.skip || 0),

    limit: Number(data.limit || params.limit),

    logs: Array.isArray(data.logs) ? data.logs : [],
  };
}

/* =====================================================
   EXPORT API
===================================================== */

export { adminUserManagementAPI };

export default adminUserManagementAPI;
